IOS-4172 Add ability to choose `ed25519` or `ed25519_slip_0010` curve

ba1ashov opened the pull request.

ba1ashov committed:
IOS-4172 Add ability to choose ed25519 or ed25519_slip_0010 curve

megakoko commented on BlockchainSdk/Common/Blockchain.swift:
```diff
@@ -667,7 +672,7 @@ extension Blockchain: Codable {
         switch key {
         case "bitcoin": self = .bitcoin(testnet: isTestnet)
-        case "stellar": self = .stellar(testnet: isTestnet)
+        case "stellar": self = .stellar(curve: curve, testnet: isTestnet)
```
А обратная совместисость с тем, что уже сохранено на диске есть?

Balashov152 replied:
ну да, получается что если сохранилась старая кривая ed25519, то она и будет

megakoko replied:
Хм, то есть если у меня сейчас есть блокчейн на главной, я его удалю и добавлю заново, то добавится он уже на другой кривой. Верно? Ничего от этого не сломается?

Первый вопрос был скорее про то, что у нас ведь сейчас параметр curve для этих блокчейнов не сохраняется. А тут мы ожидаем что он есть. Оно нормально загрузится, значение по-умолчанию правильное будет?

Balashov152 replied:
> Хм, то есть если у меня сейчас есть блокчейн на главной, я его удалю и добавлю заново, то добавится он уже на другой кривой. Верно? Ничего от этого не сломается?

не, в рамках одного конфига = карты будет одна и та же кривая

megakoko replied:
А загружаться-то нормально будет? Сейчас у .stellar не сохраняется параметр curve. А с этими изменениями мы будем его требовать. Все нормально мигрирует?

megakoko replied:
Посмотрел по коду, все норм должно быть

megakoko commented on BlockchainSdk/WalletManagers/Stellar/StellarNetworkProvider.swift:
```diff
@@ -103,7 +103,12 @@ class StellarNetworkProvider: HostProvider {
                     feeChargedP99,
                 ].map {
-                    Amount(with: blockchain, value: $0)
+                    Amount(
+                        type: .coin,
+                        currencySymbol: blockchain.currencySymbol,
+                        value: $0,
+                        decimals: blockchain.decimalCount
+                    )
```
А эти изменения нужны?

Balashov152 replied:
очень много мест, где для получения `decimalValue` нам нужен целый блокчейн создать, жутко не удобно 

Эти по факту не нужны, но надо что то делать с этим

Balashov152 replied:
Откатил

megakoko commented on BlockchainSdk/WalletManagers/Stellar/StellarNetworkProvider.swift:
```diff
@@ -20,7 +20,7 @@ class StellarNetworkProvider: HostProvider {
     
     private var blockchain: Blockchain {
-        Blockchain.stellar(testnet: isTestnet)
+        Blockchain.stellar(curve: .ed25519_slip0010, testnet: isTestnet)
```
Почему захардкожено?

Balashov152 replied:
очень много мест, где для получения decimalValue нам нужен целый блокчейн создать, жутко не удобно

tureck1y replied:
надо будет потом рефачить на decimalValue

ba1ashov committed:
IOS-4172 Add ed25519_slip0010 support

ba1ashov committed:
IOS-4172 Fix review

ba1ashov committed:
IOS-4172 Fix tests

tureck1y commented on BlockchainSdkTests/TON/TONTests.swift:
```diff
@@ -15,7 +15,7 @@ import WalletCore
 class TONTests: XCTestCase {
     
-    private var blockchain = Blockchain.ton(testnet: true)
+    private var blockchain = Blockchain.ton(curve: .ed25519_slip0010, testnet: true)
```
надо все подобные тесты на таких блокчейнах делать на двух кривых теперь

Balashov152 replied:
Дописал тесты везде где менялась кривая ✅

tureck1y commented on BlockchainSdk/Extensions/TWPublicKeyType+.swift:
```diff
@@ -26,14 +26,14 @@ extension PublicKeyType {
             case .secp256k1:
                 self = PublicKeyType.secp256k1
-            case .ed25519:
+            case .ed25519, .ed25519_slip0010:
```
интересно как это разрулить для карданы то. Получается нельзя будет это расширение использовать, может коммент напишем?

tureck1y commented on BlockchainSdk/Extensions/TWCurve+.swift:
```diff
@@ -17,7 +17,7 @@ extension Curve {
         case .secp256k1:
             self = .secp256k1
-        case .ed25519:
+        case .ed25519, .ed25519_slip0010:
```
и тут тоже по кардане надо писать коммент тогда

Balashov152 replied:
тут переделал на blockchain по аналогии с TWPublicKeyType+

tureck1y commented on BlockchainSdk/Common/Blockchain+AllCases.swift:
```diff
@@ -68,17 +68,17 @@ public extension Blockchain {
             .rsk,
```
allMainnetCases надо вообще удалять получается. Или чтобы сет id возвращался. Вот это в приложении работать не будет же

```
/// Blockchains which don't include in supported blockchains by default
    static var testable: Set<Blockchain> {
        Set(Blockchain.allMainnetCases)
            .subtracting(SupportedBlockchains(version: .v1).mainnetBlockchains())
    }
```

Balashov152 replied:
да, согласен
`allMainnetCases` убрал`public` 
Он останется только для example и для тестов

На уровне приложения получается надо будет самим добавлять в `testable: Set`, по другому что-то никак(

tureck1y commented on BlockchainSdk/Common/Blockchain.swift:
```diff
@@ -555,7 +560,7 @@ extension Blockchain {
 extension Blockchain {
     public func derivationPath(for style: DerivationStyle) -> DerivationPath? {
-        guard curve == .secp256k1 || curve == .ed25519 else {
+        guard curve == .secp256k1 || curve == .ed25519 || curve == .ed25519_slip0010 else {
```
давай сделаем через curve.supportsDerivation

Balashov152 replied:
✅

megakoko commented on BlockchainSdkTests/Common/AddressTests.swift:
```diff
@@ -342,6 +342,19 @@ class AddressesTests: XCTestCase {
         XCTAssertEqual(address.value, "rPhmKhkYoMiqC2xqHYhtPLnicWQi85uDf2") //todo: validate
     }
     
+    func testXrpEdSlip0010() throws {
+        let service = XRPAddressService(curve: .ed25519_slip0010)
+        let address = try service.makeAddress(from: edKey)
+
+        XCTAssertTrue(service.validate(address.value))
+
+        XCTAssertThrowsError(try service.makeAddress(from: secpCompressedKey))
+        XCTAssertThrowsError(try service.makeAddress(from: secpDecompressedKey))
+
+        XCTAssertEqual(address.localizedName, AddressType.default.defaultLocalizedName)
+        XCTAssertEqual(address.value, "rPhmKhkYoMiqC2xqHYhtPLnicWQi85uDf2") //todo: validate
+    }
```
Давай в функцию выше загоним чтобы копипасты не было, с циклом по кривым. Оно же на 100% так же должно работать с точки зрения блокчейна?

megakoko replied:
И с другими тоже

Balashov152 replied:
> чтобы копипасты не было

Это же тесты, она ничем не мешает, тестируем адрес сервис на разной кривой, по мне так все ок)

megakoko replied:
Можем поменять один тест и забыть обновить второй. Зачем копипаста если код на 100% одинаковый и можно без нее обойтись?

Balashov152 replied:
Можем поменять один тест и случайно заффектить другой )

ну вообще по хорошему это должны быть разные тесты, паблик ключ ed25519 и ed25519_slip0010 должны быть разные, и адреса тоже должны быть разные
кривая же разная

Balashov152 replied:
Подумаю в целом как с тестами поступить

megakoko replied:
Можно иметь единый метод, куда передавать входные-выходные данные и вызывать его из двух разных тестов

Balashov152 replied:
✅

megakoko commented on BlockchainSdkTests/VectorTests/PublicKeyDerivationAddressTests.swift:
```diff
@@ -103,9 +103,9 @@ class PublicKeyDerivationAddressTests: XCTestCase {
-            XCTFail("__INVALID_VECTOR__ \(error)")
-            return
-        }
+//        } catch {
+//            XCTFail("__INVALID_VECTOR__ \(error)")
+//            return
+//        }
```
А это что?

Balashov152 replied:
это случайно, в рамках тестов залетело

megakoko commented on BlockchainSdkTests/Vectors/blockchain_vectors.json:
```diff
@@ -14,7 +14,7 @@
         "key": "stellar",
         "testnet": false,
-        "curve": "ed25519",
+        "curve": "ed25519_slip0010",
```
Только одно будем проверять?

megakoko replied:
Так, из того что я понял нужно же и то, и то в тестах сохранить

ba1ashov committed:
IOS-4172 Add `extended:Bool` for cardano

ba1ashov committed:
IOS-4172 Fix review

megakoko commented on BlockchainSdk/WalletManagers/Polkadot/PolkadotNetwork.swift:
```diff
@@ -86,19 +87,19 @@ enum PolkadotNetwork {
 extension PolkadotNetwork {
     var existentialDeposit: Amount {
```
Чтобы на уровне приложения не было дополнительных телодвижений. Получили Amount, отформатировали как надо с currencyCode и все

Balashov152 replied:
надо подумать как работать с Amount нам, а то при изменении кривой приходится лезть туда, где по факту не меняется ничего(
тип decimalValue и currencySymbol те же, а изменения есть

ba1ashov committed:
IOS-4172 Add test ed25519_slip0010 test

IOS-4172 Update TONTests.swift

IOS-4172 Revert "IOS-4172 Fix review"

This reverts commit f486c1a0fe0de9098aba50626bba07ec6566f339.

IOS-4172 Update TONTests.swift

Balashov152 commented on BlockchainSdkTests/Solana/SolanaTests.swift:
```diff
@@ -14,6 +14,7 @@ import Combine
 
 private let raisedError = SolanaError.other("Just tx size test")
 
+// TODO: Will be refactor https://tangem.atlassian.net/browse/IOS-4260
```
Завел отдельную задачу, тут что то странное тестируется, в плане что только одна проверка на ошибку и все, больше ничего
Надо заново норм тесты написать

megakoko replied:
Мы только размер транзакции тестируем. Солана у нас работает через внешний СДК. Планируем переезд на трастволлет

Balashov152 commented on BlockchainSdkTests/TON/TONTests.swift:
```diff
@@ -32,69 +30,83 @@ class TONTests: XCTestCase {
-        } catch {
-            XCTFail("Transaction build for sign is nil")
-        }
+    func testCorrectCoinTransactionEd25519() throws {
```
не понимаю что с гитом, по факту сделал так же как и везде, у входного метода curve и два метода сверху с разными кривыми
